src/core_plugins/kibana/public/dashboard/panel/panel_header/panel_actions/build_context_menu.js

```diff
@@ -0,0 +1,112 @@
+ *     EuiContextMenuPanel,
+ *   Array.<EuiContextMenuPanelShape> childPanels - extracted child panels, if any actions also open a panel. They
+ *     need to be moved to the top level for EUI.
+ *  }}
+ */
+function buildEuiContextMenuPanelItemsAndChildPanels({ contextMenuPanelId, actions, embeddable, containerState }) {
+  const items = [];
+  const childPanels = [];
+  const actionsForPanel = getActionsForPanel(contextMenuPanelId, actions);
+  actionsForPanel.forEach(action => {
+    const isVisible = action.isVisible({ embeddable, containerState });
+    if (!isVisible) {
+      return;
+    }
+
+    let childPanelToOpenOnClick;
+    if (action.childContextMenuPanel) {
+      childPanels.push(
+        ...buildEuiContextMenuPanels({
+          contextMenuPanel: action.childContextMenuPanel,
+          actions,
+          embeddable,
+          containerState
+        }));
+      childPanelToOpenOnClick = action.childContextMenuPanel.id;
+    }
+
+    items.push(convertPanelActionToContextMenuItem(
+      {
+        action,
+        contextMenuPanelId: childPanelToOpenOnClick,
+        containerState,
+        embeddable
+      }));
+  });
+
+  return { items, childPanels };
+}
+
+/**
+ * Transforms a DashboardContextMenuPanel to the shape EuiContextMenuPanel expects, inserting any registered pluggable
+ * panel actions.
+ * @param {DashboardContextMenuPanel} contextMenuPanel
+ * @param {Array.<DashboardPanelAction>} actions to build the context menu with
+ * @param {Embeddable} embeddable
+ * @param {ContainerState} containerState
+ * @return {Object} An object that conforms to EuiContextMenuPanelShape in elastic/eui
+ */
+export function buildEuiContextMenuPanels(
+  {
+    contextMenuPanel,
+    actions,
+    embeddable,
+    containerState
+  }) {
+  const euiContextMenuPanel = {
+    id: contextMenuPanel.id,
+    title: contextMenuPanel.title,
+    items: [],
+    content: contextMenuPanel.getContent({ embeddable, containerState }),
+  };
+  let contextMenuPanels = [euiContextMenuPanel];
+
+  const { items, childPanels } =
+    buildEuiContextMenuPanelItemsAndChildPanels({
+      contextMenuPanelId: contextMenuPanel.id,
+      actions,
+      embeddable,
+      containerState
+    });
+
+  contextMenuPanels = contextMenuPanels.concat(childPanels);
+  euiContextMenuPanel.items = items;
+  return contextMenuPanels;
+}
+
+/**
+ *
+ * @param {DashboardPanelAction} action
+ * @param {String|undefined} contextMenuPanelId - an optional nested panel id to show when clicked.
+ * @param {ContainerState} containerState
+ * @param {Embeddable} embeddable
+ * @return {Object} See EuiContextMenuPanelItemShape in @elastic/eui
+ */
+function convertPanelActionToContextMenuItem({ action, contextMenuPanelId, containerState, embeddable }) {
+  return {
+    id: action.id || action.displayName.replace(/\s/g, ''),
+    name: action.displayName,
+    icon: action.icon,
+    panel: contextMenuPanelId,
```

Comment:
So, per previous comment, this would just become something like a safe get: `_.get(action, 'childContextMenuPanel.id')` or something.